File: src/content/1.8/code/scala/snippet01.scala
trait Bifunctor[F[_, _]] {
  def bimap[A, B, C, D](g: A => C)(h: B => D): F[A, B] => F[C, D] =
    first(g) compose second(h)

  def first[A, B, C](g: A => C): F[A, B] => F[C, B] =
    bimap(g)(identity[B])

  def second[A, B, D](h: B => D): F[A, B] => F[A, D] =
    bimap(identity[A])(h)
}

File: src/content/1.8/code/ocaml/snippet01.ml
module type Bifunctor = sig
  type ('a, 'b) t

  val bimap : ('a -> 'c) -> ('b -> 'd) -> ('a, 'b) t -> ('c, 'd) t
  val first : ('a -> 'c) -> ('a, 'b) t -> ('c, 'b) t
  val second : ('b -> 'd) -> ('a, 'b) t -> ('a, 'd) t
end

(** You can represent a bifunctor in one of the two forms and
    derive one from the other. *)
module type Bimap = sig
  type ('a, 'b) t

  val bimap : ('a -> 'c) -> ('b -> 'd) -> ('a, 'b) t -> ('c, 'd) t
end

module type Fmaps = sig
  type ('a, 'b) t

  val first : ('a -> 'c) -> ('a, 'b) t -> ('c, 'b) t
  val second : ('b -> 'd) -> ('a, 'b) t -> ('a, 'd) t
end

module Bifunctor_From_Bimap (M : Bimap) :
  Bifunctor with type ('a, 'b) t = ('a, 'b) M.t = struct
  include M

  let first g x = M.bimap g Fun.id x
  let second h x = M.bimap Fun.id h x
end

module Bifunctor_From_Fmaps (M : Fmaps) :
  Bifunctor with type ('a, 'b) t = ('a, 'b) M.t = struct
  include M

  let bimap g h x = M.first g (M.second h x)
end
